File: backend/internal/birdnet/clip.go
package birdnet

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os/exec"
	"path/filepath"
	"time"
)

// Clip is a stretch of a recording, written to its own file.
type Clip struct {
	// Path is where the clip is written, as a mono 16-bit WAV.
	Path string `json:"path"`
	// StartSec and EndSec are seconds into the source recording.
	StartSec float64 `json:"startSec"`
	EndSec   float64 `json:"endSec"`
}

// Recording is what Cut learned about a source file, and the clips it cut.
type Recording struct {
	DurationSec float64 `json:"durationSec"`
	SampleRate  int     `json:"sampleRate"`
	// Clips are in the order asked for, with StartSec and EndSec clamped to
	// the recording.
	Clips []Clip `json:"clips"`
}

// clipScript is analyzer/clip.py: ClipScript, or clip.py beside Script.
func (a Analyzer) clipScript() string {
	if a.ClipScript != "" {
		return a.ClipScript
	}
	return filepath.Join(filepath.Dir(a.Script), "clip.py")
}

// Cut writes each clip of source to its Path, by running analyzer/clip.py, and
// reports the source's duration and sample rate. With no clips it only reads
// those. It loads no model, so it takes about as long as Python takes to start.
func (a Analyzer) Cut(ctx context.Context, source string, clips []Clip) (Recording, error) {
	script := a.clipScript()
	if clips == nil {
		clips = []Clip{}
	}
	spec, err := json.Marshal(map[string]any{"source": source, "clips": clips})
	if err != nil {
		return Recording{}, err
	}

	var stdout bytes.Buffer
	stderr := &tailBuffer{max: 4 << 10}
	cmd := exec.CommandContext(ctx, a.Python, script)
	cmd.Stdin = bytes.NewReader(spec)
	cmd.Stdout = &stdout
	cmd.Stderr = stderr
	if a.Stderr != nil {
		cmd.Stderr = io.MultiWriter(stderr, a.Stderr)
	}
	killProcessGroupOnCancel(cmd)
	cmd.WaitDelay = 5 * time.Second

	if err := cmd.Run(); err != nil {
		if ctxErr := ctx.Err(); ctxErr != nil {
			return Recording{}, fmt.Errorf("birdnet: %w", ctxErr)
		}
		return Recording{}, fmt.Errorf("birdnet: %s: %w\n%s", script, err, stderr.String())
	}
	var rec Recording
	if err := json.Unmarshal(stdout.Bytes(), &rec); err != nil {
		return Recording{}, fmt.Errorf("birdnet: reading %s output: %w\n%s", script, err, stderr.String())
	}
	if len(rec.Clips) != len(clips) {
		return Recording{}, fmt.Errorf("birdnet: %s cut %d clips, not the %d asked for", script, len(rec.Clips), len(clips))
	}
	if rec.SampleRate <= 0 {
		return Recording{}, errors.New("birdnet: " + script + " reported no sample rate")
	}
	return rec, nil
}

File: analyzer/clip.py
"""Cut clips out of one audio file, so a detection can be heard on its own.

The Go server runs this after analyze.py (backend/internal/birdnet), once per
analyzed file, while the file is still on local disk.

    python clip.py < spec.json

stdin is one JSON object: the source file, and the clips to cut from it in
seconds into the recording.

    {"source": "/tmp/a.wav",
     "clips": [{"startSec": 731, "endSec": 739, "path": "/tmp/clip-0.flac"}]}

Each clip is written as mono 16-bit FLAC at the source's sample rate: lossless,
under half the size of the same WAV, and playable and decodable in every browser
the app supports. stdout carries exactly one JSON object: the source's length
and sample rate, and each clip as cut, clamped to the recording.

    {"durationSec": 3600.0, "sampleRate": 48000,
     "clips": [{"path": "/tmp/clip-0.flac", "startSec": 731, "endSec": 739}]}

Anything that goes wrong exits non-zero with the reason on stderr. soundfile is
what birdnet reads audio with, so any file BirdNET could analyze can be cut.
"""

import json
import sys

import soundfile


def main() -> int:
    spec = json.load(sys.stdin)
    cut = []
    with soundfile.SoundFile(spec["source"]) as src:
        rate, frames = src.samplerate, src.frames
        for clip in spec.get("clips", []):
            start = min(max(0, round(clip["startSec"] * rate)), frames)
            end = min(max(start, round(clip["endSec"] * rate)), frames)
            src.seek(start)
            audio = src.read(end - start, dtype="float32", always_2d=True)
            # Mono is what BirdNET listened to, and half the size of stereo.
            # FLAC is integer-only, so PCM_16 is the subtype either way.
            soundfile.write(clip["path"], audio.mean(axis=1), rate, format="FLAC", subtype="PCM_16")
            cut.append({
                "path": clip["path"],
                "startSec": round(start / rate, 3),
                "endSec": round((start + len(audio)) / rate, 3),
            })

    json.dump({"durationSec": round(frames / rate, 3), "sampleRate": rate, "clips": cut}, sys.stdout)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())
